pkg_resources/__init__.py

```diff
@@ -109,14 +111,18 @@
 
 
 _T = TypeVar("_T")
+_DistributionT = TypeVar("_DistributionT", bound="Distribution")
 # Type aliases
 _NestedStr = Union[str, Iterable[Union[str, Iterable["_NestedStr"]]]]
+_InstallerTypeT = Callable[["Requirement"], "_DistributionT"]
 _InstallerType = Callable[["Requirement"], Optional["Distribution"]]
 _PkgReqType = Union[str, "Requirement"]
 _EPDistType = Union["Distribution", _PkgReqType]
 _MetadataType = Optional["IResourceProvider"]
+_ResolvedEntryPoint = Any  # Can be any attribute in the module
+_ResourceStream = Any  # TODO / Incomplete: A readable file-like object
 # Any object works, but let's indicate we expect something like a module (optionally has __loader__ or __file__)
-_ModuleLike = Union[object, types.ModuleType]
+_ModuleLike = Any
```

Comment:
This annotation looks worse than the previous one, no?

Reply:
Ummm 🤔.

I would understand if `ZipProvider` was a subclass of `IResourceProvider`... however `IResourceProvider` is a protocol. That should eliminate the problem with the variance, no?

Reply:
Pyright gives us a more detailed error here:
```
Argument of type "type[ZipProvider]" cannot be assigned to parameter "provider_factory" of type "_ProviderFactoryType" in function "register_loader_type"
  Type "type[ZipProvider]" is incompatible with type "_ProviderFactoryType"
    Parameter 1: type "_ModuleLike" is incompatible with type "_ZipLoaderModule"
      Type "_ModuleLike" is incompatible with type "_ZipLoaderModule"
        "object" is incompatible with protocol "_ZipLoaderModule"
          "__loader__" is not present [reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
```

So the following might indeed be more appropriate:
```py
_ModuleLike = Union[object, types.ModuleType]
# Any: Should be _ModuleLike but we end up with variance issues
_ProviderFactoryType = Callable[[Any], "IResourceProvider"]
```

Reply:
That would also affect the `_provider_factories` variable right?

Would it make sense to add `_ZipLoaderModule` in the `_ModuleType` union?